DEVX-2410 / DEVX-2411: automate ccloud setup; add Confluent Cloud ksqlDB; DOCS-6171: screenshots

ybyzek opened the pull request:
### Description 

https://confluentinc.atlassian.net/browse/DEVX-2410
https://confluentinc.atlassian.net/browse/DEVX-2411
https://confluentinc.atlassian.net/browse/DOCS-6171


_What behavior does this PR change, and why?_

Add ksqlDB in Confluent Cloud.

Automatic way to spin up the hybrid part of the tutorial now part of the docs (analogous to `start.sh` for the original part of cp-demo)

```
./scripts/validate/validate_metrics_api.sh
```


### Author Validation

_Describe the validation already done, or needs to be done, by the PR submitter._

<!-- Uncomment any of the following that are required -->
- [x] Documentation: see staging link in https://github.com/confluentinc/docs-platform/pull/272
- [x] Run cp-demo
<!-- - [ ] jmx-monitoring-stacks -->


### Reviewer Tasks

_Describe the tasks/validation that the PR submitter is requesting to be done by the reviewer._

<!-- Uncomment any of the following that are required -->
- [ ] Documentation: see source `docs/index.rst` and staging link in https://github.com/confluentinc/docs-platform/pull/272
- [ ] Run cp-demo
<!-- - [ ] jmx-monitoring-stacks -->

mikebin commented on docs/index.rst:
```diff
@@ -1426,6 +1425,21 @@ Setup |ccloud|
 
+#. At this point, you have two options for running the |ccloud| portion of this tutorial.
+
+   The first option is to manually complete all the steps in the following sections, which we recommend if you are new to |ccloud| and |cp|: :ref:`cp-demo-ccloud-stack`, :ref:`cp-demo-telemetry-reporter`, :ref:`cp-demo-replicator-to-ccloud`, and :ref:`cp-demo-metrics-api`.
```
Should the `Cleanup` section be mentioned here?

ybyzek replied:
@mikebin agree, and fixed.  Thank you.

mikebin reviewed (approved):
One minor question, otherwise LGTM.

ybyzek commented:
@mikebin FYI, since you approved the PR, another PR has been merged into this one (it enables ksqlDB in Confluent Cloud).  The reason was to simplify validation for other folks.  Just wanted to let you know, in case you wanted to optionally review again (but not required).

tobydrake7 commented on docs/index.rst:
```diff
@@ -1426,35 +1432,47 @@ Setup |ccloud|
       ccloud login --save
 
-#. Use the :ref:`ccloud-stack` for a quick, automated way to create resources in |ccloud|.  Executed with a single command, it uses the |ccloud| CLI to:
+#. The remainder of the |ccloud| portion of this tutorial must be completed sequentially. We recommend that you manually complete all the steps in the following sections. However, you may also run the script :devx-cp-demo:`scripts/ccloud/create-ccloud-workflow.sh|scripts/ccloud/create-ccloud-workflow.sh` which automates those steps; this option is recommended for users who have run this tutorial  ...
```
```suggestion
#. The remainder of the |ccloud| portion of this tutorial must be completed sequentially. We recommend that you manually complete all the steps in the following sections. However, you may also run the script :devx-cp-demo:`scripts/ccloud/create-ccloud-workflow.sh|scripts/ccloud/create-ccloud-workflow.sh` which automates those steps. This option is recommended for users who have run this tutorial before and want to quickly bring it up.
```

tobydrake7 commented on docs/index.rst:
```diff
@@ -1426,35 +1432,47 @@ Setup |ccloud|
+-  Create a new service account.
+-  Create a new Kafka cluster and associated credentials.
+-  Enable |sr-ccloud| and associated credentials.
+-  Create ACLs with wildcard for the service account.
```
```suggestion
-  Create ACLs with a wildcard for the service account.
```

tobydrake7 commented on docs/index.rst:
```diff
@@ -1426,35 +1432,47 @@ Setup |ccloud|
+-  Create a new ksqlDB app and associated credentials
+-  Generate a local configuration file with all above connection information.
+
+#. Get a bash library of useful functions for interacting with |ccloud| (one of which is ``cloud-stack``). This library is community-supported and is not supported by Confluent.
 
```
```suggestion
#. Get a bash library of useful functions for interacting with |ccloud| (one of which is ``cloud-stack``). This library is community-supported and not supported by Confluent.

```

tobydrake7 commented on docs/index.rst:
```diff
@@ -1426,35 +1432,47 @@ Setup |ccloud|
       curl -sS -o ccloud_library.sh https://raw.githubusercontent.com/confluentinc/examples/latest/utils/ccloud_library.sh
 
-#. Using ``ccloud_library.sh`` which you downloaded in the previous step, create a new ``ccloud-stack`` (see :ref:`ccloud-stack` for advanced options). It creates real resources in |ccloud| and takes a few minutes to complete.
+#. Using ``ccloud_library.sh`` which you downloaded in the previous step, create a new ``ccloud-stack`` (see :ref:`ccloud-stack` for advanced options). It creates real resources in |ccloud| and takes a few minutes to complete. The ``true`` flag adds creation of a ksqlDB application in |ccloud|, which has hourly charges even if you are not actively using it.
```
should this be highlighted as a Caution? (charges)

ybyzek replied:
There are several other places in the docs where this is called out with more emphasis, so IMO don't think it needs to be repeated here.

tobydrake7 commented on docs/index.rst:
```diff
@@ -1487,11 +1505,14 @@ Setup |ccloud|
 ------------------
 
-#. Create a new ``Cloud`` API key and secret to authenticate with |ccloud|. These credentials will be used by the :ref:`telemetry_reporter` and used by the Metrics API, which can be used for hosted on-prem clusters as well as |ccloud| clusters.
+Enable :ref:`telemetry_reporter` on the on-prem cluster, and configure it to send metrics to the |ccloud| instance created above..
```
```suggestion
Enable :ref:`telemetry_reporter` on the on-prem cluster and configure it to send metrics to the |ccloud| instance created above..
```

tobydrake7 replied:
Since this isn't part of our standard public docs, this may not apply. However, our style is to use `on premises` instead of `on-prem` or `on prem` in docs. See https://confluentinc.atlassian.net/wiki/spaces/DOC/pages/161743785/Style+Guide+for+Confluent+Documentation. if you decide to use the approved style, do a search and replace (there are a lot of locations in the doc).

ybyzek replied:
@tobydrake7 the source is different but it all lands in docs.confluent.io, so if docs team prefers `on premises`, I can switch.  Please clarify if this guidance is merely suggested (in which case I will leave as-is) or strongly preferred (in which case I will update to `on premises`).

tobydrake7 replied:
> @tobydrake7 the source is different but it all lands in docs.confluent.io, so if docs team prefers `on premises`, I can switch. Please clarify if this guidance is merely suggested (in which case I will leave as-is) or strongly preferred (in which case I will update to `on premises`).

The doc team strongly prefers consistency and following the style guide. However, you can follow your heart on this. If you'd rather remain somewhat informal with how everyone says it (`on-prem`), in this tutorial type of document, go ahead. You know your audience for these things. If they look at `on premises` and think you're trying to use big words or something, use on-prem. 👍

ybyzek replied:
@tobydrake7 got it.  Lots of this is subjective but I'm in the `on-prem` camp, so I will leave it as is.  Thank you.

tobydrake7 commented on docs/index.rst:
```diff
@@ -1487,11 +1505,14 @@ Setup |ccloud|
-#. Create a new ``Cloud`` API key and secret to authenticate with |ccloud|. These credentials will be used by the :ref:`telemetry_reporter` and used by the Metrics API, which can be used for hosted on-prem clusters as well as |ccloud| clusters.
+Enable :ref:`telemetry_reporter` on the on-prem cluster, and configure it to send metrics to the |ccloud| instance created above..
+
+#. Create a new ``Cloud`` API key and secret to authenticate with |ccloud|. These credentials will be used by the Telemetry Reporter and used by the Metrics API, which can be used for hosted on-prem clusters as well as |ccloud| clusters.
```
```suggestion
#. Create a new ``Cloud`` API key and secret to authenticate with |ccloud|. These credentials are used by the Telemetry Reporter and the Metrics API, which can be used for hosted on-prem clusters as well as |ccloud| clusters.
```

tobydrake7 replied:
suggest: `..., which can be used for hosted` make it clear what `which` means here, does it mean the Metrics API or does it mean the API key and secret. Also, you could make this two sentences to be clearer. Put a period after Metrics API and starte the next sentence with the subject.

ybyzek replied:
Got it, I reworked this sentence to simplify it

tobydrake7 commented on docs/index.rst:
```diff
@@ -1541,22 +1562,21 @@ Telemetry Reporter
+
+Deploy |crep| to copy data from the on-prem cluster to the |ak| cluster running in |ccloud|.
+It is configured to copy from the |ak| topic ``wikipedia.parsed`` (on-prem) to the cloud topic ``wikipedia.parsed.ccloud.replica`` in |ccloud|. 
+The Replicator instance is running on the existing connect worker in the on-prem cluster.
```
```suggestion
The Replicator instance is running on the existing Connect worker in the on-prem cluster.
```

tobydrake7 replied:
always capitalize `Connect` but not connectors or worker

tobydrake7 commented on docs/index.rst:
```diff
@@ -1570,11 +1590,11 @@ Telemetry Reporter
           --connect-cluster-id connect-cluster"
 
-#. View the |crep| :devx-cp-demo:`configuration file|scripts/connectors/submit_replicator_to_ccloud_config.sh`. It is configured to copy from the |ak| topic ``wikipedia.parsed`` (on-prem) to the cloud topic ``wikipedia.parsed.ccloud.replica`` in |ccloud|. Note that it uses the local connect cluster (the origin site), so the |crep| configuration has overrides for the producer. The configuration pa ...
+#. View the |crep| :devx-cp-demo:`configuration file|scripts/connectors/submit_replicator_to_ccloud_config.sh`. Note that it uses the local connect cluster (the origin site), so the |crep| configuration has overrides for the producer. The configuration parameters that use variables are read from the env variables that you sourced in an earlier step.
```
```suggestion
#. View the |crep| :devx-cp-demo:`configuration file|scripts/connectors/submit_replicator_to_ccloud_config.sh`. Note that it uses the local connect cluster (the origin site), so the |crep| configuration has overrides for the producer. The configuration parameters that use variables are read from the environment variables you sourced in an earlier step.
```

tobydrake7 commented on docs/index.rst:
```diff
@@ -1604,9 +1626,9 @@ Metrics
       CURRENT_TIME_PLUS_1HR=$(docker-compose exec tools date -Is -d '+1 hour' | tr -d '\r')
 
-#. View the :devx-cp-demo:`metrics query file|scripts/validate/metrics_query_onprem.json`, which requests ``io.confluent.kafka.server/received_bytes`` for the topic ``wikipedia.parsed`` in the on-prem cluster (this is just one example—for examples of all the queryable metrics, see `Metrics API <https://docs.confluent.io/cloud/current/monitoring/metrics-api.html>`__).
+#. For the on-prem metrics: view the :devx-cp-demo:`metrics query file|scripts/ccloud/metrics_query_onprem.json`, which requests ``io.confluent.kafka.server/received_bytes`` for the topic ``wikipedia.parsed`` in the on-prem cluster (this is just one example—for examples of all the queryable metrics, see `Metrics API <https://docs.confluent.io/cloud/current/monitoring/metrics-api.html>`__).
```
```suggestion
#. For the on-prem metrics: view the :devx-cp-demo:`metrics query file|scripts/ccloud/metrics_query_onprem.json`, which requests ``io.confluent.kafka.server/received_bytes`` for the topic ``wikipedia.parsed`` in the on-prem cluster (for all queryable metrics examples, see `Metrics API <https://docs.confluent.io/cloud/current/monitoring/metrics-api.html>`__).
```

tobydrake7 commented on docs/index.rst:
```diff
@@ -1655,9 +1677,9 @@ Metrics
       }
 
-#. View the :devx-cp-demo:`metrics query file|scripts/validate/metrics_query_ccloud.json`, which requests ``io.confluent.kafka.server/received_bytes`` for the topic ``wikipedia.parsed.ccloud.replica`` in |ccloud| (this is just one example—for examples of all the queryable metrics, see `Metrics API <https://docs.confluent.io/cloud/current/monitoring/metrics-api.html>`__).
+#. For the |ccloud| metrics: view the :devx-cp-demo:`metrics query file|scripts/ccloud/metrics_query_ccloud.json`, which requests ``io.confluent.kafka.server/received_bytes`` for the topic ``wikipedia.parsed.ccloud.replica`` in |ccloud| (this is just one example—for examples of all the queryable metrics, see `Metrics API <https://docs.confluent.io/cloud/current/monitoring/metrics-api.html>`__).
```
```suggestion
#. For the on-prem metrics: view the :devx-cp-demo:`metrics query file|scripts/ccloud/metrics_query_onprem.json`, which requests ``io.confluent.kafka.server/received_bytes`` for the topic ``wikipedia.parsed`` in the on-prem cluster (for all queryable metrics examples, see `Metrics API <https://docs.confluent.io/cloud/current/monitoring/metrics-api.html>`__).
```

tobydrake7 replied:
```suggestion
#. For the |ccloud| metrics: view the :devx-cp-demo:`metrics query file|scripts/ccloud/metrics_query_ccloud.json`, which requests ``io.confluent.kafka.server/received_bytes`` for the topic ``wikipedia.parsed.ccloud.replica`` in |ccloud| (for all queryable metrics examples, see `Metrics API <https://docs.confluent.io/cloud/current/monitoring/metrics-api.html>`__).
```

tobydrake7 commented on docs/index.rst:
```diff
@@ -1702,12 +1724,53 @@ Metrics
+|ccloud| ksqlDB
+---------------
+
+This section creates queries in the |ccloud| ksqlDB application that processes data from the ``wikipedia.parsed.ccloud.replica`` topic that |crep| copied from the on-prem cluster.
+You must have completed :ref:`cp-demo-ccloud-stack` before proceeding.
```
```suggestion
This section shows how to create queries in the |ccloud| ksqlDB application that processes data from the ``wikipedia.parsed.ccloud.replica`` topic that |crep| copied from the on-prem cluster.
You must have completed :ref:`cp-demo-ccloud-stack` before proceeding.
```

tobydrake7 commented on docs/index.rst:
```diff
@@ -1702,12 +1724,53 @@ Metrics
+
+   .. figure:: images/ccloud_ksqldb_stream.png
+
+#. The ksqlDB application in |ccloud| has hourly charges even if you are not actively using it, so after you are finished, proceed to :ref:`cp-demo-ccloud-cleanup`.
```
```suggestion
#. Go to :ref:`cp-demo-ccloud-cleanup` to clean up the demo. Important: The ksqlDB application in |ccloud| has hourly charges even if you are not actively using it.
```

tobydrake7 reviewed (approved):
LGTM minor comments

michaelpeacock commented:
1. I created a cloud account, but the instructions did not state to create a cluster. This step is where you enter your credit card which is needed in later steps (running ccloud::create_ccloud_stack).

2. ccloud::create_ccloud_stack would exit before being able to read error (no credit) and kill the terminal. I modified the function in the script to do a return instead of exit.

3. docker-compose logs kafka1 | grep confluent.telemetry.api.key doesn't return secret as shown in the output. (I removed .key from the grep)

4. CURRENT_TIME_MINUS_1HR=$(docker-compose exec tools date -Is -d '-1 hour' | tr -d '\r')
CURRENT_TIME_PLUS_1HR=$(docker-compose exec tools date -Is -d '+1 hour' | tr -d '\r')  --> not copied from copy option

5. I could not run the script in Confluent Cloud ksqlDB (Step 4) to create the kstreams below. I believe it is just a formatting issue. I can the commands individually and they worked. Error below:
{"@type":"generic_error","error_code":40000,"message":"Invalid JSON in request: Cannot construct instance of `java.util.LinkedHashMap` (although at least one Creator exists): no String-argument constructor/factory method to deserialize from String value ('ksql.streams.auto.offset.reset')\n at [Source: (byte[])\"{  \"ksql\": \"CREATE STREAM wikipedia WITH (kafka_topic='wikipedia.parsed.ccloud.replica', value_format='AVRO');\",  \"streamsProperties\": \"ksql.streams.auto.offset.reset\":\"earliest\"}\"; line: 1, column: 136] (through reference chain: io.confluent.ksql.rest.entity.KsqlRequest[\"streamsProperties\"])"}{"@type":"generic_error","error_code":40000,"message":"Invalid JSON in request: Cannot construct instance of `java.util.LinkedHashMap` (although at least one Creator exists): no String-argument constructor/factory method to deserialize from String value ('ksql.streams.auto.offset.reset')\n at [Source: (byte[])\"{  \"ksql\": \"CREATE STREAM wikipedianobot AS SELECT *, (length->new - length->old) AS BYTECHANGE FROM wikipedia WHERE bot = false AND length IS NOT NULL AND length->new IS NOT NULL AND length->old IS NOT NULL;\",  \"streamsProperties\": \"ksql.streams.auto.offset.reset\":\"earliest\"}\"; line: 1, column: 234] (through reference chain: io.confluent.ksql.rest.entity.KsqlRequest[\"streamsProperties\"])"}

**These worked fined:**
curl -X "POST" "$KSQLDB_ENDPOINT/ksql" \
     -H "Content-Type: application/vnd.ksql.v1+json; charset=utf-8" \
     -u $KSQLDB_BASIC_AUTH_USER_INFO \
     --silent \
     -d $'{
  "ksql": "CREATE STREAM wikipedia WITH (kafka_topic=\'wikipedia.parsed.ccloud.replica\', value_format=\'AVRO\');",
  "streamsProperties": {"ksql.streams.auto.offset.reset":"earliest","ksql.streams.cache.max.bytes.buffering":"0"}
}'

curl -X "POST" "$KSQLDB_ENDPOINT/ksql" \
     -H "Content-Type: application/vnd.ksql.v1+json; charset=utf-8" \
     -u $KSQLDB_BASIC_AUTH_USER_INFO \
     --silent \
     -d $'{
  "ksql": "CREATE STREAM wikipedianobot AS SELECT *, (length->new - length->old) AS BYTECHANGE FROM wikipedia WHERE bot = false AND length IS NOT NULL AND length->new IS NOT NULL AND length->old IS NOT NULL;",
  "streamsProperties": {"ksql.streams.auto.offset.reset":"earliest","ksql.streams.cache.max.bytes.buffering":"0"}
}'

6. Noticeable issues with both brokers after running for about a 1/2 hour. Needed to restart the containers multiple times. Docker memory was increased to 8GB prior to starting as recommended.

ybyzek commented:
@michaelpeacock thanks for the feedback.  Addressing points below:

1. Latest commit calls out adding a payment method.
2. ccloud_stack_create to not exit terminal --> tracking in https://confluentinc.atlassian.net/browse/DEVX-2426
3. This does work in my environment.  Is it possible there is something else happening in your environment?
4. Both get copied on my machine.  Could it be something else happening?  What OS are you using?
5. Sometimes what gets copied into the clipboard differs from the original text.  Is it possible for you to show what got copied into the clipboard?  (copy/paste here). For example, maybe the environment is adding unnecessary backslashes.
6. Longer cp-demo runs --> tracking in https://confluentinc.atlassian.net/browse/DEVX-2427. I have run cp-demo for at least 30 minutes without issue in my environment, but this jira will be more intentional about observing resource utilization and try to determine what might be happening

michaelpeacock commented:
@ybyzek 

3. The line below does not get displayed as shown in the output since you are grepping for confluent.telemetry.api.key and there is no key in the output.
kafka1            |       confluent.telemetry.api.secret = [hidden]

4. Worked for me today. Must have been operator error.

ybyzek commented:
@michaelpeacock I originally misunderstood #3 issue.  I now understand the issue you raised and am fixing in the docs in the latest output, thank you.
